tests/pools/rai/integration/test_redemption_rate_handling.py

```diff
@@ -0,0 +1,237 @@
+
+        amount = int(10 ** self.decimals[send] * st_pct)
+        value = amount if self.underlying_coins[send] == ETH_ADDRESS else 0
+        self.swap.exchange(send, recv, amount, 0, {"from": self.alice, "value": value})
+
+    def rule_generate_fees(self):
+        """
+        Pushes base pool to be heavily imbalanced and then rebalances it to generate a lot of fees
+        and thereby increase the virtual price.
+        """
+        min_idx, max_idx = self._min_max_underlying(base=True)
+        dx = self.base_swap.balances(max_idx)
+        base_decimals = self.underlying_decimals[self.n_coins - 1:]
+        if base_decimals[max_idx] > base_decimals[min_idx]:
+            dx = dx / 10 ** (base_decimals[max_idx] - base_decimals[min_idx])
+        elif base_decimals[min_idx] > base_decimals[max_idx]:
+            dx = dx * 10 ** (base_decimals[min_idx] - base_decimals[max_idx])
+
+        base_coins = self.underlying_coins[self.n_coins - 1:]
+        base_coins[min_idx]._mint_for_testing(self.alice, dx, {"from": self.alice})
+
+        tx = self.base_swap.exchange(min_idx, max_idx, dx, 0, {"from": self.alice})
+        dy = tx.events["TokenExchange"]["tokens_bought"]
+        self.base_swap.exchange(max_idx, min_idx, dy, 0, {"from": self.alice})
+
+    def rule_exchange_underlying(self, st_pct):
+        """
+        Perform a swap using underlying coins.
+        """
+        send, recv = self._min_max_underlying()
+
+        amount = int(10 ** self.underlying_decimals[send] * st_pct)
+        value = amount if self.underlying_coins[send] == ETH_ADDRESS else 0
+        self.swap.exchange_underlying(send, recv, amount, 0, {"from": self.alice, "value": value})
+
+    def rule_remove_one_coin(self, st_pct):
+        """
+        Remove liquidity from the pool in only one coin.
+        """
+        if not hasattr(self.swap, "remove_liquidity_one_coin"):
+            # not all pools include `remove_liquidity_one_coin`
+            return self.rule_remove_imbalance(st_pct)
+
+        idx = self._min_max()[1]
+        amount = int(10 ** self.decimals[idx] * st_pct)
+        self.swap.remove_liquidity_one_coin(amount, idx, 0, {"from": self.alice})
+
+    def rule_remove_imbalance(self, st_pct):
+        """
+        Remove liquidity from the pool in an imbalanced manner.
+        """
+        idx = self._min_max()[1]
+        amounts = [0] * self.n_coins
+        amounts[idx] = 10 ** self.decimals[idx] * st_pct
+        self.swap.remove_liquidity_imbalance(amounts, 2 ** 256 - 1, {"from": self.alice})
+
+    def rule_remove(self, st_pct):
+        """
+        Remove liquidity from the pool.
+        """
+        amount = int(10 ** 18 * st_pct)
+        self.swap.remove_liquidity(amount, [0] * self.n_coins, {"from": self.alice})
+
+    def rule_adjust_redemption_price(self, st_red_prices):
+        self.redemption_price_snap.setRedemptionPriceSnap(st_red_prices)
+
+    def invariant_check_virtual_price(self):
```

Comment:
This checks only base virtual price. It will never go down, that's checked in other places. But the actual swap virtual price could!

And this makes it hard to test. Price going down inevitably leads to a loss, but in principle price going up should lead to virtual_price going up: maybe that's something which can be tested.